
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      
      
      <link rel="icon" href="assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.29">
    
    
      
        <title>PanMAN Wiki</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/main.76a95c52.min.css">
      
        
        <link rel="stylesheet" href="assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="white" data-md-color-accent="white">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#welcome-to-panman-wiki" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="index.html" title="PanMAN Wiki" class="md-header__button md-logo" aria-label="PanMAN Wiki" data-md-component="logo">
      
  <img src="images/icon.svg" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            PanMAN Wiki
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Home
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="white" data-md-color-accent="white"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="white" data-md-color-accent="white"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7 0-.24-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91 1.61 0 2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08Z"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/TurakhiaLab/panman" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    TurakhiaLab/panman
  </div>
</a>
      </div>
    
  </nav>
  
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



  

<nav class="md-nav md-nav--primary md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="index.html" title="PanMAN Wiki" class="md-nav__button md-logo" aria-label="PanMAN Wiki" data-md-component="logo">
      
  <img src="images/icon.svg" alt="logo">

    </a>
    PanMAN Wiki
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/TurakhiaLab/panman" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    TurakhiaLab/panman
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Home
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="index.html" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#what-are-panmans" class="md-nav__link">
    <span class="md-ellipsis">
      What are PanMANs?
    </span>
  </a>
  
    <nav class="md-nav" aria-label="What are PanMANs?">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#panmans-protocol-buffer-file-format" class="md-nav__link">
    <span class="md-ellipsis">
      PanMAN's Protocol Buffer file format
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#panmanutils" class="md-nav__link">
    <span class="md-ellipsis">
      panmanUtils
    </span>
  </a>
  
    <nav class="md-nav" aria-label="panmanUtils">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#panmanutils-video-tutorial" class="md-nav__link">
    <span class="md-ellipsis">
      panmanUtils Video Tutorial
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    <span class="md-ellipsis">
      Installation
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Installation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#installation-scripts" class="md-nav__link">
    <span class="md-ellipsis">
      Installation Scripts
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#docker" class="md-nav__link">
    <span class="md-ellipsis">
      Docker
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#construction-of-panmans-using-panmanutils" class="md-nav__link">
    <span class="md-ellipsis">
      Construction of PanMANs using panmanUtils
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#functionalities-in-panmanutils" class="md-nav__link">
    <span class="md-ellipsis">
      Functionalities in panmanUtils
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Functionalities in panmanUtils">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#summary-extract" class="md-nav__link">
    <span class="md-ellipsis">
      Summary extract
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#newick-extract" class="md-nav__link">
    <span class="md-ellipsis">
      Newick extract
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#extended-newick-extract" class="md-nav__link">
    <span class="md-ellipsis">
      Extended Newick extract
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tipinternal-node-sequences-extract" class="md-nav__link">
    <span class="md-ellipsis">
      Tip/internal node sequences extract
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#multiple-sequence-alignment-msa-extract" class="md-nav__link">
    <span class="md-ellipsis">
      Multiple Sequence Alignment (MSA) extract
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#multiple-whole-genome-alignment-m-wga-extract" class="md-nav__link">
    <span class="md-ellipsis">
      Multiple Whole Genome Alignment (m-WGA) extract
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#variant-call-format-vcf-extract" class="md-nav__link">
    <span class="md-ellipsis">
      Variant Call Format (VCF) extract
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#graphical-fragment-assembly-gfa-extract" class="md-nav__link">
    <span class="md-ellipsis">
      Graphical fragment assembly (GFA) extract
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#subnetwork-extract" class="md-nav__link">
    <span class="md-ellipsis">
      Subnetwork extract
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#annotate" class="md-nav__link">
    <span class="md-ellipsis">
      Annotate
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#amino-acid-translation" class="md-nav__link">
    <span class="md-ellipsis">
      Amino Acid Translation
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contributions" class="md-nav__link">
    <span class="md-ellipsis">
      Contributions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#citing-panman" class="md-nav__link">
    <span class="md-ellipsis">
      Citing PanMAN
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
    <a href="https://github.com/TurakhiaLab/panman/edit/master/docs/index.md" title="Edit this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4v-2m10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1 2.1 2.1Z"/></svg>
    </a>
  
  
    
      
    
    <a href="https://github.com/TurakhiaLab/panman/raw/master/docs/index.md" title="View source of this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 18c.56 0 1 .44 1 1s-.44 1-1 1-1-.44-1-1 .44-1 1-1m0-3c-2.73 0-5.06 1.66-6 4 .94 2.34 3.27 4 6 4s5.06-1.66 6-4c-.94-2.34-3.27-4-6-4m0 6.5a2.5 2.5 0 0 1-2.5-2.5 2.5 2.5 0 0 1 2.5-2.5 2.5 2.5 0 0 1 2.5 2.5 2.5 2.5 0 0 1-2.5 2.5M9.27 20H6V4h7v5h5v4.07c.7.08 1.36.25 2 .49V8l-6-6H6a2 2 0 0 0-2 2v16a2 2 0 0 0 2 2h4.5a8.15 8.15 0 0 1-1.23-2Z"/></svg>
    </a>
  


<h1 id="welcome-to-panman-wiki">Welcome to PanMAN Wiki</h1>
<div align="center">
    <img src="images/logo.svg"/>
</div>

<h2 id="what-are-panmans"><b>What are PanMANs?</b></h2>
<p>PanMAN or Pangenome Mutation-Annotated Network is a novel data representation for pangenomes that provides massive leaps in both representative power and storage efficiency. Specifically, PanMANs are composed of mutation-annotated trees, called PanMATs, which, in addition to substitutions, also annotate inferred indels (Fig. 2b), and even structural mutations (Fig. 2a) on the different branches. Multiple PanMATs are connected in the form of a network using edges to generate a PanMAN (Fig. 2c). PanMAN's representative power is compared against existing pangenomic formats in Fig. 1. PanMANs are the most compressible pangenomic format for the different microbial datasets (SARS-CoV-2, RSV, HIV, Mycobacterium. Tuberculosis, E. Coli, and Klebsiella pneumoniae), providing 2.9 to 559-fold compression over standard pangenomic formats. </p>
<div align="center">
    <img src="images/representpower.svg" width="600" height="600"/><br>
    <b>Figure 1: Comparison of representative power of PanMAN against other pangenomic formats (yellow ticks indicate partial representative ability)</b><br>
</div>
<p><br>
<br></p>
<div align="center">
    <img src="images/panman.svg" width="600" height="600"/><br>
    <b>Figure 2: Overview of the PanMAN data structure</b><br>
</div>

<h3 id="panmans-protocol-buffer-file-format"><b>PanMAN's Protocol Buffer file format</b></h3>
<p>PanMAN utilizes Google’s protocol buffer (protobuf, <a href="https://protobuf.dev/">https://protobuf.dev/</a>), a binary serialization file format, to compactly store PanMAN's data structure in a file. Fig. 3 provides the .proto file defining the PanMAN’s structure. At the top level, the file format of PanMANs encodes a list (declared as a repeated identifier in the .protof file) of PanMATs. Each PanMAT object stores the following data elements: (a) a unique identifier, (b) a phylogenetic tree stored as a string in Newick format, (c) a list of mutations on each branch ordered according to the pre-order traversal of the tree topology, (d) a block mapping object to record homologous segments identified as duplications and rearrangements, which are mapped against their common consensus sequence; the block-mapping object is also used to derive the pseudo-root, e) a gap list to store the position and length of gaps corresponding to each block's consensus sequence. Each mutation object encodes the node's block and nucleotide mutations that are inferred on the branches leading to that node. If a block mutation exists at a position described by the Block-ID field (int32), the block mutation field (bool) is set to 1, otherwise set to 0, and its type is stored as a substitution to and from a gap in Block mutation type field (bool), encoded as 0 or 1, respectively. In PanMAN, each nucleotide mutation within a block inferred on a branch has four pieces of information, i.e., position (middle coordinate), gap position (last coordinate), mutation type, and mutated characters. To reduce redundancy in the file, consecutive mutations of the same type are packed together and stored as a mutation info (int32) field, where mutation type, mutation length, and mutated characters use 3, 5, and 24 bits, respectively. PanMAN stores each character using one-hot encoding, hence, one "Nucleotide Mutations" object can store up to 6 consecutive mutations of the same type. PanMAN's file also stores the complex mutation object to encode the type of complex mutation and its metadata such as PanMATs' and nodes' identifiers, breakpoint coordinates, etc. The entire file is then compressed using XZ (<a href="https://github.com/tukaani-project/xz">https://github.com/tukaani-project/xz</a>) to enhance storage efficiency.</p>
<div align="center">
    <img src="images/pb.svg" width="600" height="600"/><br>
    <b>Figure 3: PanMAN's file format</b>
</div>

<h2 id="panmanutils"><i><b>panmanUtils</b></i></h2>
<p><i>panmanUtils</i> includes multiple algorithms to construct PanMANs and to support various functionalities to modify and extract useful information from PanMANs (Fig. 4). </p>
<div align="center">
    <img src="images/utility.svg" width="600" height="600"/><br>
    <b>Figure 4:  Overview of panmanUtils' functionalities</b>
</div>

<h3 id="panmanutils-video-tutorial"><b><i>panmanUtils</i> Video Tutorial</b></h3>
<p>TBA</p>
<h3 id="installation"><b>Installation</b></h3>
<p>panmanUtils can be installed using two different options, as described below: <br>
1. Installation script <br>
2. Docker</p>
<h4 id="installation-scripts">Installation Scripts</h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>git clone https://github.com/TurakhiaLab/panman.git
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>cd panman/install
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>./installUbuntu.sh
</code></pre></div>
<h4 id="docker">Docker</h4>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>docker run -it swalia14/panman:latest
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a># Inside the docker container
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>git clone https://github.com/TurakhiaLab/panman.git
<a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a>cd panman/install
<a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a>./installUbuntu.sh
</code></pre></div>
<h3 id="construction-of-panmans-using-panmanutils"><b>Construction of PanMANs using <i>panmanUtils</i></b></h3>
<p>Since PanMAN can be composed of a single or multiple PanMATs, to construct a PanMAN, we start with a single tree PanMAN (or PanMAT) and then split it up into a network of multiple PanMATs using the inferred complex mutations provided as input. To construct the starting single-tree PanMAN representing a collection of sequences, panmanUtils require two inputs:<br>
1. Tree topology representing the phylogenetic relationship of the input sequences <br>
2. A pangenomic data structure - PanGraph(Recommended, see below)/GFA/FASTA, representing the multiple-sequence alignment (MSA) corresponding to the sequence collection.</p>
<ul>
<li>Example syntax and Usage<br>
Construct PanGraph (JSON format) and tree topology (Newick format) from raw genome sequences using PanGraph tool
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>$ ./pangraph build -k mmseqs -a α -b β &lt;path to input raw sequences in FASTA format&gt; | ./pangraph polish
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>$ ./pangraph build -k mmseqs -a α -b β ecoli_10.fa | ./pangraph polish
</code></pre></div>
Construct single tree PanMAN using PanGraph as input
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>$ ./panmanUtils --pangraph-in=&lt;path to PanGraph JSON file&gt; --newick-in=&lt;path to newick file&gt; --output-file=&lt;prefix of panman&#39;s file name&gt;
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>$ ./panmanUtils --pangraph-in=ecoli_10.json --newick-in=ecoli_10.nwk --output-file=ecoli_10
</code></pre></div>
Similarly, if you'd like to construct a PanMAN using a GFA or an MSA as an input, use the <code>--gfa-in</code> or the <code>--msa-in</code> options instead of <code>--pangraph-in</code>.<blockquote>
<p><strong>NOTE:</strong> Currently, we only support GFAv1.1 consisting of Segments, un-overlapping Links and Paths.</p>
</blockquote>
</li>
</ul>
<h3 id="functionalities-in-panmanutils"><b>Functionalities in <i>panmanUtils</i></b></h3>
<p>All panmanUtils functionality commands manipulate the input PanMAN file.
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>$ ./panmanUtils -I &lt;path to PanMAN file&gt; {opt}
</code></pre></div></p>
<p>Specific options:
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>  -I [ --input-panman ] arg   Input PanMAN file path
<a id="__codelineno-7-2" name="__codelineno-7-2" href="#__codelineno-7-2"></a>  -P [ --input-pangraph ] arg Input PanGraph JSON file to build a PanMAN
<a id="__codelineno-7-3" name="__codelineno-7-3" href="#__codelineno-7-3"></a>  -G [ --input-gfa ] arg      Input GFA file to build a PanMAN
<a id="__codelineno-7-4" name="__codelineno-7-4" href="#__codelineno-7-4"></a>  -M [ --input-msa ] arg      Input MSA file (FASTA format) to build a PanMAN
<a id="__codelineno-7-5" name="__codelineno-7-5" href="#__codelineno-7-5"></a>  -N [ --input-newick ] arg   Input tree topology as Newick string
<a id="__codelineno-7-6" name="__codelineno-7-6" href="#__codelineno-7-6"></a>
<a id="__codelineno-7-7" name="__codelineno-7-7" href="#__codelineno-7-7"></a>  -s [ --summary ]            Print PanMAN summary
<a id="__codelineno-7-8" name="__codelineno-7-8" href="#__codelineno-7-8"></a>  -t [ --newick ]             Print newick string of all trees in a PanMAN
<a id="__codelineno-7-9" name="__codelineno-7-9" href="#__codelineno-7-9"></a>  -f [ --fasta ]              Print tip/internal sequences (FASTA format)
<a id="__codelineno-7-10" name="__codelineno-7-10" href="#__codelineno-7-10"></a>  -m [ --fasta-aligned ]      Print MSA of sequences for each PanMAT in a PanMAN (FASTA format)
<a id="__codelineno-7-11" name="__codelineno-7-11" href="#__codelineno-7-11"></a>  -b [ --subnet ]             Extract subnet of given PanMAN to a new PanMAN 
<a id="__codelineno-7-12" name="__codelineno-7-12" href="#__codelineno-7-12"></a>                              file based on the list of nodes provided in the 
<a id="__codelineno-7-13" name="__codelineno-7-13" href="#__codelineno-7-13"></a>                              input-file
<a id="__codelineno-7-14" name="__codelineno-7-14" href="#__codelineno-7-14"></a>  -v [ --vcf ]                Print variations of all sequences from any PanMAT
<a id="__codelineno-7-15" name="__codelineno-7-15" href="#__codelineno-7-15"></a>                              in a PanMAN (VCF format)
<a id="__codelineno-7-16" name="__codelineno-7-16" href="#__codelineno-7-16"></a>  -g [ --gfa ]                Convert any PanMAT in a PanMAN to a GFA file
<a id="__codelineno-7-17" name="__codelineno-7-17" href="#__codelineno-7-17"></a>  -w [ --maf ]                Print m-WGA for each PanMAT in a PanMAN (MAF 
<a id="__codelineno-7-18" name="__codelineno-7-18" href="#__codelineno-7-18"></a>                              format)
<a id="__codelineno-7-19" name="__codelineno-7-19" href="#__codelineno-7-19"></a>  -a [ --annotate ]           Annotate nodes of the input PanMAN based on the 
<a id="__codelineno-7-20" name="__codelineno-7-20" href="#__codelineno-7-20"></a>                              list provided in the input-file
<a id="__codelineno-7-21" name="__codelineno-7-21" href="#__codelineno-7-21"></a>  -r [ --reroot ]             Reroot a PanMAT in a PanMAN based on the input 
<a id="__codelineno-7-22" name="__codelineno-7-22" href="#__codelineno-7-22"></a>                              sequence id (--reference)
<a id="__codelineno-7-23" name="__codelineno-7-23" href="#__codelineno-7-23"></a>  -v [ --aa-translation ]     Extract amino acid translations in tsv file
<a id="__codelineno-7-24" name="__codelineno-7-24" href="#__codelineno-7-24"></a>  -e [ --extended-newick ]    Print PanMAN&#39;s network in extended-newick format
<a id="__codelineno-7-25" name="__codelineno-7-25" href="#__codelineno-7-25"></a>  -k [ --create-network ]     Create PanMAN with network of trees from single 
<a id="__codelineno-7-26" name="__codelineno-7-26" href="#__codelineno-7-26"></a>                              or multiple PanMAN files
<a id="__codelineno-7-27" name="__codelineno-7-27" href="#__codelineno-7-27"></a>  -p [ --printMutations ]     Create PanMAN with network of trees from single 
<a id="__codelineno-7-28" name="__codelineno-7-28" href="#__codelineno-7-28"></a>                              or multiple PanMAN files
<a id="__codelineno-7-29" name="__codelineno-7-29" href="#__codelineno-7-29"></a>
<a id="__codelineno-7-30" name="__codelineno-7-30" href="#__codelineno-7-30"></a>  -q [ --acr ]                ACR method [fitch(default), mppa]
<a id="__codelineno-7-31" name="__codelineno-7-31" href="#__codelineno-7-31"></a>  -n [ --reference ] arg      Identifier of reference sequence for PanMAN 
<a id="__codelineno-7-32" name="__codelineno-7-32" href="#__codelineno-7-32"></a>                              construction (optional), VCF extract (required), 
<a id="__codelineno-7-33" name="__codelineno-7-33" href="#__codelineno-7-33"></a>                              or reroot (required)
<a id="__codelineno-7-34" name="__codelineno-7-34" href="#__codelineno-7-34"></a>  -s [ --start ] arg          Start coordinate of protein translation
<a id="__codelineno-7-35" name="__codelineno-7-35" href="#__codelineno-7-35"></a>  -e [ --end ] arg            End coordinate of protein translation
<a id="__codelineno-7-36" name="__codelineno-7-36" href="#__codelineno-7-36"></a>  -d [ --treeID ] arg         Tree ID, required for --vcf
<a id="__codelineno-7-37" name="__codelineno-7-37" href="#__codelineno-7-37"></a>  -i [ --input-file ] arg     Path to the input file, required for --subnet, 
<a id="__codelineno-7-38" name="__codelineno-7-38" href="#__codelineno-7-38"></a>                              --annotate, and --create-network
<a id="__codelineno-7-39" name="__codelineno-7-39" href="#__codelineno-7-39"></a>  -o [ --output-file ] arg    Prefix of the output file name
</code></pre></div></p>
<blockquote>
<p><strong>NOTE:</strong> When output-file argument is optional and is not provided to <i>panmanUtils</i>, the output will be printed in the terminal.</p>
</blockquote>
<h4 id="summary-extract">Summary extract</h4>
<p>The summary feature extracts node and tree level statistics of a PanMAN, that contains a summary of its geometric and parsimony information.</p>
<ul>
<li>Example syntax and Usage
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>$ ./panmanUtils -I &lt;path to PanMAN file&gt; --summary --output-file=&lt;prefix of output file&gt; (optional)
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>$ ./panmanUtils -I ecoli_10.panman --summary --output-file=ecoli_10
</code></pre></div></li>
</ul>
<h4 id="newick-extract">Newick extract</h4>
<p>Extract Newick string of all trees in a PanMAN.</p>
<ul>
<li>Example syntax and Usage
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>$ ./panmanUtils -I &lt;path to PanMAN file&gt; --newick --output-file=&lt;prefix of output file&gt; (optional)
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>$ ./panmanUtils -I ecoli_10.panman --newick --output-file=ecoli_10
</code></pre></div></li>
</ul>
<h4 id="extended-newick-extract">Extended Newick extract</h4>
<p>Extract network in Extended Newick format.</p>
<ul>
<li>Example syntax and Usage
<div class="highlight"><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a>$ ./panmanUtils -I &lt;path to PanMAN file&gt; ----extended-newick --output-file=&lt;prefix of output file&gt; (optional)
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a>$ ./panmanUtils -I ecoli_10.panman ----extended-newick --output-file=ecoli_10
</code></pre></div></li>
</ul>
<h4 id="tipinternal-node-sequences-extract">Tip/internal node sequences extract</h4>
<p>Extract tip and internal node sequences from a PanMAN in a FASTA format.</p>
<ul>
<li>Example syntax and Usage
<div class="highlight"><pre><span></span><code><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a>$ ./panmanUtils -I &lt;path to PanMAN file&gt; --fasta --output-file=&lt;prefix of output file&gt; (optional)
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a>$ ./panmanUtils -I ecoli_10.panman --fasta --output-file=ecoli_10
</code></pre></div></li>
</ul>
<h4 id="multiple-sequence-alignment-msa-extract">Multiple Sequence Alignment (MSA) extract</h4>
<p>Extract MSA of sequences for each PanMAT (with pseduo-root coordinates) in a PanMAN in a FASTA format.</p>
<ul>
<li>Example syntax and Usage
<div class="highlight"><pre><span></span><code><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a>$ ./panmanUtils -I &lt;path to PanMAN file&gt; --fasta-aligned --output-file=&lt;prefix of output file&gt; (optional)
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a>$ ./panmanUtils -I ecoli_10.panman --fasta-aligned --output-file=ecoli_10
</code></pre></div></li>
</ul>
<h4 id="multiple-whole-genome-alignment-m-wga-extract">Multiple Whole Genome Alignment (m-WGA) extract</h4>
<p>Extract m-WGA for each PanMAT in a PanMAN in the form of a UCSC multiple alignment format (MAF).</p>
<ul>
<li>Example syntax and Usage
<div class="highlight"><pre><span></span><code><a id="__codelineno-18-1" name="__codelineno-18-1" href="#__codelineno-18-1"></a>$ ./panmanUtils -I &lt;path to PanMAN file&gt; --maf --output-file=&lt;prefix of output file&gt; (optional)
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-19-1" name="__codelineno-19-1" href="#__codelineno-19-1"></a>$ ./panmanUtils -I ecoli_10.panman --maf --output-file=ecoli_10
</code></pre></div></li>
</ul>
<h4 id="variant-call-format-vcf-extract">Variant Call Format (VCF) extract</h4>
<p>Extract variations of all sequences from any PanMAT in a PanMAN in the form of a VCF file with respect to <i>any</i> reference sequence (ref) in the PanMAT.</p>
<ul>
<li>Example syntax and Usage
<div class="highlight"><pre><span></span><code><a id="__codelineno-20-1" name="__codelineno-20-1" href="#__codelineno-20-1"></a>$ ./panmanUtils -I &lt;path to PanMAN file&gt; --vcf -reference=ref --output-file=&lt;prefix of output file&gt; (optional) 
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-21-1" name="__codelineno-21-1" href="#__codelineno-21-1"></a>$ ./panmanUtils -I ecoli_10.panman --vcf -reference=NC_000913.3 --output-file=ecoli_10 
</code></pre></div></li>
</ul>
<h4 id="graphical-fragment-assembly-gfa-extract">Graphical fragment assembly (GFA) extract</h4>
<p>Convert any PanMAT in a PanMAN to a Graphical fragment assembly (GFA) file representing the pangenome.</p>
<ul>
<li>Example syntax and Usage
<div class="highlight"><pre><span></span><code><a id="__codelineno-22-1" name="__codelineno-22-1" href="#__codelineno-22-1"></a>$ ./panmanUtils -I &lt;path to PanMAN file&gt; --gfa --output-file=&lt;prefix of output file&gt; (optional)
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-23-1" name="__codelineno-23-1" href="#__codelineno-23-1"></a>$ ./panmanUtils -I ecoli_10.panman --gfa --output-file=ecoli_10 
</code></pre></div></li>
</ul>
<h4 id="subnetwork-extract">Subnetwork extract</h4>
<p>Extract a subnetwork from a given PanMAN and write it to a new PanMAN file based on the list of nodes provided in the input-file.</p>
<ul>
<li>Example syntax and Usage
<div class="highlight"><pre><span></span><code><a id="__codelineno-24-1" name="__codelineno-24-1" href="#__codelineno-24-1"></a>$ ./panmanUtils -I &lt;path to PanMAN file&gt; --subnet --input-file=&lt;path to a file containing list of nodes&gt; --output-file=&lt;prefix of output file&gt;
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-25-1" name="__codelineno-25-1" href="#__codelineno-25-1"></a>$ ./panmanUtils -I ecoli_10.panman --subnet --input-file=nodes.txt --output-file=ecoli_10_subnet
</code></pre></div></li>
</ul>
<h4 id="annotate">Annotate</h4>
<p>Annotate nodes in a PanMAN with a custom string, later searched by these annotations, using an input TSV file containing a list of nodes and their corresponding custom annotations. </p>
<ul>
<li>Example syntax and Usage
<div class="highlight"><pre><span></span><code><a id="__codelineno-26-1" name="__codelineno-26-1" href="#__codelineno-26-1"></a>$ ./panmanUtils -I &lt;path to PanMAN file&gt; --annotate &lt;path to file containing list of annotations&gt; --output-file=ecoli_10_annotate
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-27-1" name="__codelineno-27-1" href="#__codelineno-27-1"></a>$ ./panmanUtils -I ecoli_10.panman --annotate --input-file=annotations.tsv --output-file=ecoli_10_annotate
</code></pre></div><blockquote>
<p><strong>NOTE:</strong> If output-file is not provided to <i>panmanUtils</i>, the annotated PanMAN will be written to the same file.</p>
</blockquote>
</li>
</ul>
<h4 id="amino-acid-translation">Amino Acid Translation</h4>
<p>Extract amino acid translations from a PanMAN in TSV file.</p>
<ul>
<li>Example syntax and Usage
<div class="highlight"><pre><span></span><code><a id="__codelineno-28-1" name="__codelineno-28-1" href="#__codelineno-28-1"></a>$ ./panmanUtils -I &lt;path to PanMAN file&gt; --aa-translations --output-file=&lt;prefix of output file&gt; (optional)
</code></pre></div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-29-1" name="__codelineno-29-1" href="#__codelineno-29-1"></a>$ ./panmanUtils -I ecoli_10.panman --aa-translations --output_file=ecoli_10
</code></pre></div></li>
</ul>
<h2 id="contributions"><b>Contributions</b></h2>
<p>We welcome contributions from the community to enhance the capabilities of PanMAN and panmanUtils. If you encounter any issues or have suggestions for improvement, please open an issue on <a href="https://github.com/TurakhiaLab/panman">PanMAN GitHub page</a>. For general inquiries and support, reach out to our team.</p>
<h2 id="citing-panman"><b>Citing PanMAN</b></h2>
<p>If you use the PanMANs or panmanUtils in your research or publications, we kindly request that you cite the following paper:<br> 
* Sumit Walia, Harsh Motwani, Kyle Smith, Russell Corbett-Detig, Yatish Turakhia, "<i>Compressive Pangenomics Using Mutation-Annotated Networks</i>", bioRxiv 2024.07.02.601807; doi: <a href="https://doi.org/10.1101/2024.07.02.601807">10.1101/2024.07.02.601807</a></p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
    
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/TurakhiaLab/panman" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": ".", "features": ["announce.dismiss", "content.action.edit", "content.action.view", "content.code.annotate", "content.code.copy", "content.tooltips", "navigation.footer", "navigation.expand", "navigation.tabs.sticky", "navigation.instant.prefetch", "navigation.tracking", "search.highlight", "search.share", "search.suggest", "toc.follow", "toc.integrate"], "search": "assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="assets/javascripts/bundle.fe8b6f2b.min.js"></script>
      
        <script src="javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>